### PR TITLE
add tianshou-like JAX+PPO+Mujoco

### PR DESCRIPTION
**Pull request by quangr**
## Description
Add  tianshou-like JAX+PPO+Mujoco code, which is tested in Anv-v3 and HalfCheetah-v3.

11 seed test

Anv-v3  (**Tianshou 1M:3258.4+-1079.3 ; 3M:4079.3+-880.2**)
my result:
Anv-v3

HalfCheetah-v3 (**Tianshou 1M:5783.9+-1244.0 ; 3M:7337.4+-1508.2**)
my result:
HalfCheetah-v3

I guess the difference performence in Anv-v3 is because in tianshou they will discard the frame returnded from truncated state and reset. But in jax there is no easy way to discard these frame, we take these trash frame and count reset as a step.

 Ant-v3 is a relatively short length env, so the replaybuffer will contain a lot of these useless frames.

## Types of changes

- [ ] Bug fix
- [x] New feature
- [ ] New algorithm
- [ ] Documentation

## Checklist:

- [x ] I've read the [CONTRIBUTION](https://github.com/vwxyzjn/cleanrl/blob/master/CONTRIBUTING.md) guide (**required**).
- [x] I have ensured `pre-commit run --all-files` passes (**required**).
- [x] I have updated the documentation and previewed the changes via `mkdocs serve`.
- [ ] I have updated the tests accordingly (if applicable).

If you are adding new algorithm variants or your change could result in performance difference, you may need to (re-)run tracked experiments. See https://github.com/vwxyzjn/cleanrl/pull/137 as an example PR. 
- [ ] I have contacted [vwxyzjn](https://github.com/vwxyzjn) to obtain access to the [openrlbenchmark W&B team](https://wandb.ai/openrlbenchmark) (**required**).
- [ ] I have tracked applicable experiments in [openrlbenchmark/cleanrl](https://wandb.ai/openrlbenchmark/cleanrl) with `--capture-video` flag toggled on (**required**).
- [ ] I have added additional documentation and previewed the changes via `mkdocs serve`.
    - [ ] I have explained note-worthy implementation details.
    - [ ] I have explained the logged metrics.
    - [ ] I have added links to the original paper and related papers (if applicable).
    - [ ] I have added links to the PR related to the algorithm variant.
    - [ ] I have created a table comparing my results against those from reputable sources (i.e., the original paper or other reference implementation).
    - [ ] I have added the learning curves (in PNG format).
    - [ ] I have added links to the tracked experiments.
    - [ ] I have updated the overview sections at the [docs](https://docs.cleanrl.dev/rl-algorithms/overview/) and the [repo](https://github.com/vwxyzjn/cleanrl#overview)
- [ ] I have updated the tests accordingly (if applicable).

### CODE REVIEW

File: cleanrl/ppo_mujoco_envpool_xla_jax.py [hunk 1]
```diff
@@ -0,0 +1,679 @@
...
+                values[1:] * (1.0 - dones[1:]),
+                values[:-1],
+                storage.rewards,
+            ),
+            reverse=True,
+        )
+        returns = advantages + values[:-1]
+        if args.rew_norm:
+            returns = (returns / jnp.sqrt(agent_state.ret_rms.var + 10e-8)).astype(jnp.float32)
+            agent_state = agent_state.replace(ret_rms=agent_state.ret_rms.update(returns.flatten()))
```
Comment (vwxyzjn) [inline cleanrl/ppo_mujoco_envpool_xla_jax.py:450]: Interesting. Going through the logic here I realize this reward normalization is quite different from the original implementation, which may or may not be fine but it's worth pointing out the difference.

The original implementation does a "forward discounted return" and normalizes the reward on a per-step basis. 
image
See [source1](https://github.com/openai/baselines/blob/ea25b9e8b234e6ee1bca43083f8f3cf974143998/baselines/common/vec_env/vec_normalize.py#L28) and [source2](https://iclr-blog-track.github.io/2022/03/25/ppo-implementation-details/)

Here what happens is that the returns are normalized only after the rollout phase, so the rewards themselves are not normalized.


It doesn't feel like these two approaches are equivalent. If you want to get to the bottom of this, it might be worth it to conduct an empirical study comparing this implementation and https://github.com/vwxyzjn/cleanrl/pull/348.

Both approaches seem fine to me, but we should document if we choose the current apporach.

File: cleanrl/ppo_mujoco_envpool_xla_jax.py [hunk 2]
```diff
@@ -0,0 +1,679 @@
...
+            entropy_loss,
+            approx_kl,
+            grads,
+        ) = jax.lax.scan(update_epoch, (agent_state, key), (), length=args.update_epochs)
+        return agent_state, loss, pg_loss, v_loss, entropy_loss, approx_kl, key
+
+    # TRY NOT TO MODIFY: start the game
+    global_step = 0
+    start_time = time.time()
+    handle, next_obs = envs.reset()
```
Comment (vwxyzjn) [inline cleanrl/ppo_mujoco_envpool_xla_jax.py:556]: Is this an API change...? I thought `handle, recv, send, step_env = envs.xla()` already gave you the handle.

File: cleanrl/ppo_mujoco_envpool_xla_jax.py [hunk 3]
```diff
@@ -0,0 +1,679 @@
+# docs and experiment results can be found at https://docs.cleanrl.dev/rl-algorithms/ppo/#ppo_atari_envpool_xla_jaxpy
```
Comment (vwxyzjn) [inline cleanrl/ppo_mujoco_envpool_xla_jax.py:1]: I suggest renaming the file to `ppo_continuous_action_envpool_xla_jax_scan.py` to be consistent with the naming convention of other variants. We might also be able to apply this file on the DMControl envs from envpool (https://envpool.readthedocs.io/en/latest/env/dm_control.html)

**Review by vwxyzjn — CHANGES_REQUESTED** [review]
Hi @quangr, thanks for this awesome contribution! Being able to use JAX+PPO+MuJoCo+EnvPool will be a game-changer for a lot of folks! This PR will also make #217 not necessary.

Some comments and thoughts:
* Would you mind sharing your wandb username so I can add you to the `openrlbenchmark` entity? It would be great if you could contribute tracked experiments there, and we can use our CLI utility (https://github.com/openrlbenchmark/openrlbenchmark) to plot charts.
* ~~Could you share your huggingface username and help add saved models?~~
    * On a second thought there might not be a way to render mujoco images with Envpool. Don't worry about this yet.
    * ~~We recently added the huggingface integration as follows https://github.com/vwxyzjn/cleanrl/blob/d0d6baed3910e6aebc0d5aba9c7cd10267d19e57/cleanrl/ppo_atari_envpool_xla_jax_scan.py#L477-L511~~
    * ~~You can load the trained model by running `python -m cleanrl_utils.enjoy --exp-name ppo_atari_envpool_xla_jax_scan --env-id Breakout-v5`~~
    * image

File: cleanrl/ppo_mujoco_envpool_xla_jax.py [hunk 4]
```diff
@@ -0,0 +1,679 @@
...
+        var = jnp.exp(2 * stdlog)
+        logprob = (-((action - logits) ** 2) / (2 * var) - stdlog - jnp.log(jnp.sqrt(2 * jnp.pi))).sum(-1)
+        entropy = (2 * stdlog + jnp.log(2 * jnp.pi) + 1) / 2
+        value = critic.apply(params.critic_params, x).squeeze()
+        return logprob, entropy.sum(-1), value
+
+    def compute_gae_once(carry, inp, gamma, gae_lambda):
+        advantages = carry
+        nextdone, nexttruncated, nextvalues, curvalues, reward = inp
+        nextnonterminal = (1.0 - nextdone) * (1.0 - nexttruncated)
```
[the remark this answers is not in the document]
Comment (51616) [reply cleanrl/ppo_mujoco_envpool_xla_jax.py:410]: Just curious. Are we currently using gymnasium? My last PR still uses the normal gym interface. This the version bump a dependency of any of these libs?

File: cleanrl/ppo_mujoco_envpool_xla_jax.py [hunk 5]
```diff
@@ -0,0 +1,679 @@
...
+        global_step += args.num_steps * args.num_envs
+        agent_state, loss, pg_loss, v_loss, entropy_loss, approx_kl, key = update_ppo(
+            agent_state,
+            next_obs,
+            next_done,
+            next_truncated,
+            storage,
+            key,
+        )
+        print(f"pg_loss={pg_loss.mean()}, loss={loss.mean()}, v_loss={v_loss.mean()}, entropy_loss={entropy_loss.mean()}")
```
Comment (51616) [inline cleanrl/ppo_mujoco_envpool_xla_jax.py:657]: Please move the logging down. From the reference `ppo_continuous_action.py`, `pg_loss` should come after `v_loss`

File: cleanrl/ppo_mujoco_envpool_xla_jax.py [hunk 6]
```diff
@@ -0,0 +1,679 @@
...
+
+        # TRY NOT TO MODIFY: record rewards for plotting purposes
+        writer.add_scalar("charts/avg_episodic_return", avg_episodic_return, global_step)
+        writer.add_scalar("charts/avg_episodic_length", avg_episodic_length, global_step)
+        writer.add_scalar("charts/learning_rate", agent_state.opt_state[1].hyperparams["learning_rate"].item(), global_step)
+        writer.add_scalar("losses/value_loss", v_loss.mean().item(), global_step)
+        writer.add_scalar("losses/policy_loss", pg_loss.mean().item(), global_step)
+        writer.add_scalar("losses/entropy", entropy_loss.mean().item(), global_step)
+        writer.add_scalar("losses/approx_kl", approx_kl.mean().item(), global_step)
+        writer.add_scalar("losses/loss", loss.mean().item(), global_step)
```
Comment (51616) [inline cleanrl/ppo_mujoco_envpool_xla_jax.py:672]: All the logged values should use the value from the last minibatch as opposed to the mean (please see `ppo_atari_envpool_xla_jax_scan.py`). This deviation might cause some confusion as to why the logged values are not consistent with previous implementations.

**Review by 51616 — COMMENTED** [review]
Thank you for a nice PR! Still, there are some unjustified changes which might cause performance difference vs other versions. In general, aim to reduce the number of code changes (see https://github.com/vwxyzjn/cleanrl/pull/328#issuecomment-1341481576 for how to compare your pr with a reference implementation). For your case, it might be a good idea to compare to `ppo_continuous_action.py` and `ppo_atari_envpool_xla_jax_scan.py`. A good rule of thumb is to stay close to these as possible.

File: cleanrl/ppo_mujoco_envpool_xla_jax.py (hunk 1 above)
Comment (vwxyzjn) [reply cleanrl/ppo_mujoco_envpool_xla_jax.py:450]: I also noticed there is no observation normalization. Is this intended?

**Comment by quangr** [comment]
> 你好@quangr，感谢您的贡献！对于很多人来说，能够使用 JAX+PPO+MuJoCo+EnvPool 将改变游戏规则！此 PR 也将使#217变得不必要。
> 
> 一些评论和想法：
> 
> * 您介意分享您的 wandb 用户名以便我将您添加到该`openrlbenchmark`实体吗？如果您可以在那里贡献跟踪实验，那就太好了，我们可以使用我们的 CLI 实用程序 ( https://github.com/openrlbenchmark/openrlbenchmark ) 来绘制图表。

Thank @51616 and @vwxyzjn  for code reviewing[❤️](https://emojipedia.org/red-heart/)! My wandb username is [quangr](https://wandb.ai/quangr). I will check these comment and improve on my code soon.

**Comment by quangr** [comment]
I have run experiment for Ant-v4 HalfCheetah-v4  Hopper-v4  Walker2d-v4  Swimmer-v4 Humanoid-v4 Reacher-v4 InvertedPendulum-v4 InvertedDoublePendulum-v4. Here is the report https://wandb.ai/openrlbenchmark/cleanrl/reports/MuJoCo-jax-EnvPool--VmlldzozNDczNDkz.

# comparing the result to tianshou

I notice that tianshou use 10 envs to evaluate performance from reset state every epoch as their benchmark, I wonder if it's a problem for comparing.

# comparing with ppo_continuous_action_8M
better : Ant-v4 HalfCheetah-v4 
similar :Hopper-v4 Walker2d-v4 Humanoid-v4  Reacher-v4
worse :Swimmer-v4 
In tianshou benchmark, their parameter also act poorly in Swimmer. So it's a sensonable result.

as for InvertedDoublePendulum-v4 InvertedPendulum-v4, every agents in my version reach 1000 score, which not happens in ppo_continuous_action_8M. But it start to decline afterward, and in tainshou's training data we can observe same decline curve: https://drive.google.com/drive/folders/1tQvgmsBbuLPNU3qo5thTBi03QzGXygXf
https://drive.google.com/drive/folders/1ns2cGnAn_39wqCItmhDZIxihLi8-DBei

**Comment by vwxyzjn** [comment]
Thanks for running the results! They look great. Feel free to click `resolve conversation` as you resolve the PR comments and let me know when it's ready for another review.

Meanwhile, you might find the following tool handy
```
pip install openrlbenchmark
python -m openrlbenchmark.rlops \
    --filters '?we=openrlbenchmark&wpn=cleanrl&ceik=env_id&cen=exp_name&metric=charts/episodic_return' 'ppo_continuous_action_8M?tag=v1.0.0-13-gcbd83f6'  \
    --filters '?we=openrlbenchmark&wpn=cleanrl&ceik=env_id&cen=exp_name&metric=charts/avg_episodic_return' 'ppo_continuous_action_envpool_xla_jax_scan?tag=v1.0.0-jax-ca-be3113b'   \
    --env-ids Ant-v4 InvertedDoublePendulum-v4 Reacher-v4  Hopper-v4 HalfCheetah-v4 Swimmer-v4  Humanoid-v4 InvertedPendulum-v4 Walker2d-v4 \
    --check-empty-runs False \
    --ncols 4 \
    --ncols-legend 3 \
    --output-filename ppo_continuous_action_envpool_xla_jax_scan \
    --scan-history --report
```

Which generates 

ppo_continuous_action_envpool_xla_jax_scan
ppo_continuous_action_envpool_xla_jax_scan-time

and the wandb report:

https://wandb.ai/costa-huang/cleanRL/reports/Regression-Report-ppo_continuous_action_envpool_xla_jax_scan--VmlldzozNDgzMzM4


Couple of notes:
* Would you mind running the experiments for `HumanoidStandup-v4, Pusher-v4` as well?
* FWIW, it is possible to get high scores in `Humanoid-v4` as well. 

See 

```
python -m openrlbenchmark.rlops \
    --filters '?we=openrlbenchmark&wpn=cleanrl&ceik=env_id&cen=exp_name&metric=charts/episodic_return' 'ppo_continuous_action_8M?tag=v1.0.0-13-gcbd83f6'  \
    --filters '?we=openrlbenchmark&wpn=cleanrl&ceik=env_id&cen=exp_name&metric=charts/avg_episodic_return' 'ppo_continuous_action_envpool_xla_jax_scan?tag=v1.0.0-jax-ca-be3113b'   \
    --filters '?we=openrlbenchmark&wpn=envpool-cleanrl&ceik=env_id&cen=exp_name&metric=charts/episodic_return' 'ppo_continuous_action_envpool' \
    --env-ids HalfCheetah-v4 Walker2d-v4 Hopper-v4 InvertedPendulum-v4 Humanoid-v4  \
    --check-empty-runs False \
    --ncols 4 \
    --ncols-legend 3 \
    --output-filename ppo_continuous_action_envpool_xla_jax_scan \
    --scan-history
```

ppo_continuous_action_envpool_xla_jax_scan

The hyperparams used is in https://github.com/vwxyzjn/envpool-cleanrl/blob/880552a168c08af334b5e5d8868bfbe5ea881445/ppo_continuous_action_envpool.py#L40-L75, with no obs nor reward normalization.

**Comment by quangr** [comment]
> Thanks for running the results! They look great. Feel free to click `resolve conversation` as you resolve the PR comments and let me know when it's ready for another review.
> 
> Meanwhile, you might find the following tool handy
> 
> ```
> pip install openrlbenchmark
> python -m openrlbenchmark.rlops \
>     --filters '?we=openrlbenchmark&wpn=cleanrl&ceik=env_id&cen=exp_name&metric=charts/episodic_return' 'ppo_continuous_action_8M?tag=v1.0.0-13-gcbd83f6'  \
>     --filters '?we=openrlbenchmark&wpn=cleanrl&ceik=env_id&cen=exp_name&metric=charts/avg_episodic_return' 'ppo_continuous_action_envpool_xla_jax_scan?tag=v1.0.0-jax-ca-be3113b'   \
>     --env-ids Ant-v4 InvertedDoublePendulum-v4 Reacher-v4  Hopper-v4 HalfCheetah-v4 Swimmer-v4  Humanoid-v4 InvertedPendulum-v4 Walker2d-v4 \
>     --check-empty-runs False \
>     --ncols 4 \
>     --ncols-legend 3 \
>     --output-filename ppo_continuous_action_envpool_xla_jax_scan \
>     --scan-history --report
> ```
> 
> Which generates
> 
> ppo_continuous_action_envpool_xla_jax_scan ppo_continuous_action_envpool_xla_jax_scan-time
> 
> and the wandb report:
> 
> https://wandb.ai/costa-huang/cleanRL/reports/Regression-Report-ppo_continuous_action_envpool_xla_jax_scan--VmlldzozNDgzMzM4
> 
> Couple of notes:
> 
> * Would you mind running the experiments for `HumanoidStandup-v4, Pusher-v4` as well?
> * FWIW, it is possible to get high scores in `Humanoid-v4` as well.
> 
> See
> 
> ```
> python -m openrlbenchmark.rlops \
>     --filters '?we=openrlbenchmark&wpn=cleanrl&ceik=env_id&cen=exp_name&metric=charts/episodic_return' 'ppo_continuous_action_8M?tag=v1.0.0-13-gcbd83f6'  \
>     --filters '?we=openrlbenchmark&wpn=cleanrl&ceik=env_id&cen=exp_name&metric=charts/avg_episodic_return' 'ppo_continuous_action_envpool_xla_jax_scan?tag=v1.0.0-jax-ca-be3113b'   \
>     --filters '?we=openrlbenchmark&wpn=envpool-cleanrl&ceik=env_id&cen=exp_name&metric=charts/episodic_return' 'ppo_continuous_action_envpool' \
>     --env-ids HalfCheetah-v4 Walker2d-v4 Hopper-v4 InvertedPendulum-v4 Humanoid-v4  \
>     --check-empty-runs False \
>     --ncols 4 \
>     --ncols-legend 3 \
>     --output-filename ppo_continuous_action_envpool_xla_jax_scan \
>     --scan-history
> ```
> 
> ppo_continuous_action_envpool_xla_jax_scan
> 
> The hyperparams used is in https://github.com/vwxyzjn/envpool-cleanrl/blob/880552a168c08af334b5e5d8868bfbe5ea881445/ppo_continuous_action_envpool.py#L40-L75, with no obs nor reward normalization.

Thanks for updating me. I'll be ready for the code review once the documentation is finished. I'm also happy to run the experiment you suggested.

**Comment by quangr** [comment]
I have document the questions you brought up and I am now ready for the code review.  I would be happy to hear your feedback.
